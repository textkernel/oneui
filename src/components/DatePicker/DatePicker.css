/* Input field */
.react-datepicker__input-container input {
    height: 30px;
    width: 230px;
    border: var(--color-neutral) 1px solid;
    border-radius: var(--border-radius);
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.1);
    padding: var(--spacing-normal) var(--spacing-normal) var(--spacing-normal) var(--spacing-3x);
}

.react-datepicker__input-container input:focus,
.react-datepicker__input-container input:focus-visible {
    outline: none;
    border-color: var(--color-brand);
}

/* Calendar container */
.react-datepicker-popper[data-placement^='bottom'] {
    padding-top: var(--spacing-2x);
}
.react-datepicker-popper[data-placement^='top'] {
    padding-bottom: var(--spacing-2x);
}

.react-datepicker {
    border-radius: var(--border-radius);
    border: var(--color-neutral) 1px solid;
    box-shadow: 0 5px 10px 0 rgba(0, 0, 0, 0.2);
    padding: var(--spacing-3x);
}

/* Header area */

.react-datepicker__header {
    background-color: var(--color-background);
    box-sizing: border-box;
    padding: 0;
    border-bottom: 0;
}

.react-datepicker__day-names {
    margin-bottom: 0;
}
.react-datepicker__day-name {
    color: var(--color-neutral-75);
    width: 35px;
    margin: 0;
}

.react-datepicker__month {
    margin: 0;
}

/* Days */

.react-datepicker__week {
    border: var(--color-neutral-25) 1px solid;
}

.react-datepicker__week ~ .react-datepicker__week {
    border-top: 0;
}

.react-datepicker__day {
    margin: 0;
    width: 35px;
    height: 35px;
    font-size: var(--font-size-normal);
    padding: var(--spacing-normal);
    box-sizing: border-box;
    color: var(--color-neutral-75);
}

.react-datepicker__day ~ .react-datepicker__day {
    border-left: var(--color-neutral-25) 1px solid;
}

.react-datepicker__day:hover {
    border-radius: 0;
    background-color: var(--color-neutral-25);
    color: var(--color-foreground);
}

.react-datepicker__day--outside-month {
    color: var(--color-neutral);
}

.react-datepicker__day--keyboard-selected {
    border-radius: 0;
    background-color: var(--color-neutral-25);
    color: var(--color-foreground);
}

.react-datepicker__day--keyboard-selected:hover {
    background-color: var(--color-neutral-25);
}

/* Selected days */

.react-datepicker__day--selected {
    border-radius: 0;
    background-color: var(--color-brand);
    color: var(--color-background);
}

.react-datepicker__day--in-selecting-range,
.react-datepicker__day--in-selecting-range:not(.react-datepicker__day--in-range, .react-datepicker__month-text--in-range, .react-datepicker__quarter-text--in-range, .react-datepicker__year-text--in-range),
.react-datepicker__day--in-range {
    border-radius: 0;
    background-color: var(--color-brand-40);
    color: var(--color-background);
}

.react-datepicker__day--selecting-range-start,
.react-datepicker__day--selecting-range-end {
    border-radius: 0;
    background-color: var(--color-brand);
    color: var(--color-background);
}

/* Today button */
.react-datepicker__today-button {
    background-color: var(--color-background);
    border: 0;
    color: var(--color-brand);
    font-weight: var(--font-weight-normal);
    padding: var(--spacing-3x) 0;
}
