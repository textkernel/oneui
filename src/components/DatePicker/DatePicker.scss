@import '../../themes/oneui/oneui.scss';

/* Input field */
:global .react-datepicker__input-container input {
    @extend %input;

    &:focus,
    &:focus-visible {
        outline: none;
        border-color: var(--color-primary);
    }
}

/* Calendar container */
:global .react-datepicker-popper[data-placement^='bottom'] {
    padding-top: var(--spacing-2x);
}
:global .react-datepicker-popper[data-placement^='top'] {
    padding-bottom: var(--spacing-2x);
}

:global .react-datepicker {
    font-family: var(--font-family-primary);
    border-radius: var(--border-radius);
    border: var(--color-light) 1px solid;
    box-shadow: 0 5px 10px 0 rgba(0, 0, 0, 0.2);
    padding: var(--spacing-3x);
}

/* Header area */

:global .react-datepicker__header {
    background-color: var(--color-background);
    box-sizing: border-box;
    padding: 0;
    border-bottom: 0;
}

:global .react-datepicker__day-names {
    margin-bottom: 0;
}
:global .react-datepicker__day-name {
    color: var(--color-neutral-75);
    width: 35px;
    margin: 0;
}

:global .react-datepicker__month {
    margin: 0;
}

/* Days */

:global .react-datepicker__week {
    border: var(--color-light-25) 1px solid;
}

:global .react-datepicker__week ~ .react-datepicker__week {
    border-top: 0;
}

:global .react-datepicker__day {
    margin: 0;
    width: 35px;
    height: 35px;
    font-size: var(--font-size-normal);
    padding: var(--spacing-normal);
    box-sizing: border-box;
    color: var(--color-neutral-75);

    & ~ .react-datepicker__day {
        border-left: var(--color-light-25) 1px solid;
    }

    &:hover {
        border-radius: 0;
        background-color: var(--color-light-25);
        color: var(--color-neutral-75);
    }
}

:global .react-datepicker__day--outside-month {
    color: var(--color-neutral);
}

:global .react-datepicker__day--keyboard-selected {
    border-radius: 0;
    background-color: var(--color-light);
    color: var(--color-neutral-75);

    &:hover {
        background-color: var(--color-neutral-25);
    }
}

:global .react-datepicker__day--disabled {
    color: var(--color-neutral-25);

    &:hover {
        background-color: var(--color-background);
        color: var(--color-neutral-40);
    }
}
/* Selected days */

:global .react-datepicker__day--selected {
    border-radius: 0;
    background-color: var(--color-primary);
    color: var(--color-background);
}

/* stylelint-disable */
:global .react-datepicker__day--in-selecting-range,
:global
    .react-datepicker__day--in-selecting-range:not(.react-datepicker__month-text--in-range, .react-datepicker__quarter-text--in-range, .react-datepicker__year-text--in-range),
:global .react-datepicker__day--in-range {
    border-radius: 0;
    background-color: var(--color-primary-40);
    color: var(--color-background);
}
/* stylelint-enable */

:global .react-datepicker__day--selecting-range-start,
:global .react-datepicker__day--selecting-range-end {
    border-radius: 0;
    background-color: var(--color-primary);
    color: var(--color-background);
}

/* Today button */
:global .react-datepicker__today-button {
    background-color: var(--color-background);
    padding: var(--spacing-2x) var(--spacing-4x) 0 var(--spacing-4x);
    border: 0;
}
