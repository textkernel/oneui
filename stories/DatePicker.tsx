import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { select, text, withKnobs } from '@storybook/addon-knobs';
import { DatePicker, Text } from '@textkernel/oneui';
import es from 'date-fns/locale/es';
import hu from 'date-fns/locale/hu';
import frCA from 'date-fns/locale/fr-CA';
import { registerLocale } from 'react-datepicker';

const DATE_REGEX = /\d\d\d\d-\d\d-\d\d/;

storiesOf('Molecules|DatePicker', module)
    .addDecorator(withKnobs)
    .add(
        'DatePicker',
        () => {
            registerLocale('es', es);
            registerLocale('hu', hu);
            registerLocale('fr-CA', frCA);

            const [selected, setSelected] = React.useState(new Date());

            const handleChange = (date: Date) => {
                setSelected(date);
                console.log('new date selected: ', date);
            };

            const minDateStr = text('Min date yyyy-MM-dd', '');
            const maxDateStr = text('Max date yyyy-MM-dd', '');

            let minDate;
            if (minDateStr.match(DATE_REGEX)) {
                minDate = new Date(minDateStr);
            }
            let maxDate;
            if (maxDateStr.match(DATE_REGEX)) {
                maxDate = new Date(maxDateStr);
            }

            return (
                <DatePicker
                    onChange={handleChange}
                    selected={selected}
                    minDate={minDate}
                    maxDate={maxDate}
                    todayButton="Today"
                    locale={select(
                        'Locale (see code example or lib docs for implementation details)',
                        ['en', 'es', 'hu', 'fr-CA'],
                        'en'
                    )}
                />
            );
        },
        {
            info: {
                text: `
            ## Usage information

            This component is a thin wrapper around [react-datepicker](https://github.com/Hacker0x01/react-datepicker/), 
            only setting the header element and add css.

            You can pass [props](https://github.com/Hacker0x01/react-datepicker/blob/master/docs/datepicker.md) according to their definition.
            See also [examples](https://reactdatepicker.com/) on their website. 
            `,
            },
        }
    )
    .add(
        'DateRangePicker',
        () => {
            const [startDate, setStartDate] = React.useState(new Date());
            const [endDate, setEndDate] = React.useState<Date | null>(null);

            const minDateStr = text('Min date yyyy-MM-dd', '');
            const maxDateStr = text('Max date yyyy-MM-dd', '');

            let minDate;
            if (minDateStr.match(DATE_REGEX)) {
                minDate = new Date(minDateStr);
            }
            let maxDate;
            if (maxDateStr.match(DATE_REGEX)) {
                maxDate = new Date(maxDateStr);
            }

            const handleStartChange = (date: Date) => {
                setStartDate(date);
                if (endDate && date > endDate) {
                    setEndDate(null);
                }
                document.getElementById('range-end')?.focus();
            };

            const handleEndChange = (date: Date) => {
                setEndDate(date);
            };

            return (
                <>
                    <Text>Start date:</Text>
                    <DatePicker
                        id="range-start"
                        selected={startDate}
                        onChange={handleStartChange}
                        selectsStart
                        startDate={startDate}
                        endDate={endDate}
                        minDate={minDate}
                        maxDate={maxDate}
                    />
                    <Text>End date:</Text>
                    <DatePicker
                        id="range-end"
                        selected={endDate}
                        onChange={handleEndChange}
                        selectsEnd
                        startDate={startDate}
                        endDate={endDate}
                        minDate={startDate}
                        maxDate={maxDate}
                    />
                </>
            );
        },
        {
            info: {
                text: `
    ## Example implementation

    This is an example on how to compose a range date picker using DatePicker component.
    OneUI does not provide a ready component for it.

    __Note__: DatePicker is a thin wrapper around [react-datepicker](https://github.com/Hacker0x01/react-datepicker/). 
    You can pass [props](https://github.com/Hacker0x01/react-datepicker/blob/master/docs/datepicker.md) according to their definition.
    See also [examples](https://reactdatepicker.com/) on their website.
    `,
            },
        }
    );
