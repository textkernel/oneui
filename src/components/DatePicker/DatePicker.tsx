import * as React from 'react';
import ReactDatePicker, { ReactDatePickerProps } from 'react-datepicker';
import { CalendarHeader } from './CalendarHeader';
import 'react-datepicker/dist/react-datepicker.css?external'; // eslint-disable-line import/no-unresolved
import './DatePicker.scss';

const DEFAULT_YEAR_RANGE = 100;

export const DatePicker: React.FC<ReactDatePickerProps> = (props) => {
    const { children, minDate, maxDate, locale, ...rest } = props;

    const currentYear = new Date().getFullYear();
    const defaultMinDate = new Date();
    defaultMinDate.setFullYear(currentYear - DEFAULT_YEAR_RANGE);
    const defaultMaxDate = new Date();
    defaultMaxDate.setFullYear(currentYear + DEFAULT_YEAR_RANGE);

    const minYear = (minDate || defaultMinDate).getFullYear();
    const maxYear = (maxDate || defaultMaxDate).getFullYear();

    let localeStr;
    if (locale) {
        if (typeof locale === 'string') {
            localeStr = locale;
        } else {
            localeStr = locale.code;
        }
    }

    return (
        <ReactDatePicker
            {...rest}
            showPopperArrow={false}
            locale={locale}
            minDate={minDate || defaultMinDate}
            maxDate={maxDate || defaultMaxDate}
            renderCustomHeader={(propsFromLib) => (
                <CalendarHeader
                    {...propsFromLib}
                    locale={localeStr}
                    yearsRange={[minYear, maxYear]}
                />
            )}
            autoComplete="off"
        >
            {children}
        </ReactDatePicker>
    );
};

DatePicker.displayName = 'DatePicker';
