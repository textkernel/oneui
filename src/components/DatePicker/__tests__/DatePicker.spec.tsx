import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import MockDate from 'mockdate';
import { DatePicker } from '../DatePicker';

describe('DatePicker', () => {
    let view: RenderResult;
    const handleChange = jest.fn();

    const minDate = new Date('2021-11-23');
    const maxDate = new Date('2023-03-04');

    const openCalendar = async (user) => {
        await user.click(screen.getByRole('textbox'));
    };

    beforeAll(() => {
        MockDate.set('2021-12-24T13:00:00.000Z');
    });

    afterAll(() => {
        MockDate.reset();
    });

    beforeEach(() => {
        view = render(<DatePicker onChange={handleChange} minDate={minDate} maxDate={maxDate} />);
    });

    it('should render correctly', () => {
        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('textbox')).toBeInTheDocument();
    });

    it('should render calendar header correctly when calendar is open', async () => {
        const user = userEvent.setup();
        await openCalendar(user);

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('presentation')).toBeInTheDocument();
        expect(screen.getAllByRole('option')).toHaveLength(38);
        expect(screen.getByRole('presentation')).toBeInTheDocument();
    });

    it('should set min and max date on calendar even if they are no originally passed as props', async () => {
        view.rerender(<DatePicker onChange={handleChange} />);
        const user = userEvent.setup();
        // this is to make sure the calendar navigation and the year selection in the header are always aligned
        await openCalendar(user);

        expect(
            screen.getByRole('option', { name: minDate.getFullYear().toString() })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('option', { name: maxDate.getFullYear().toString() })
        ).toBeInTheDocument();
    });

    it('should call onChange function with Date object', async () => {
        const user = userEvent.setup();
        // A simplistic regression test for the library
        await user.type(screen.getByRole('textbox'), '11/1/2022');

        expect(handleChange).toHaveBeenCalledTimes(1);
        expect(handleChange.mock.calls[0][0] instanceof Date).toBeTruthy();
    });

    it('should set year boundaries on header when minDate and maxDate are passed', async () => {
        const user = userEvent.setup();

        await openCalendar(user);

        expect(screen.getByText(minDate.getFullYear())).toBeInTheDocument();
        expect(screen.getByText(maxDate.getFullYear())).toBeInTheDocument();
    });

    describe('when no min/max date props passed', () => {
        beforeEach(() => {
            view.rerender(
                <DatePicker onChange={handleChange} minDate={undefined} maxDate={undefined} />
            );
        });

        it('should set min/max date to default', async () => {
            const user = userEvent.setup();

            await openCalendar(user);

            expect(view.container).toMatchSnapshot();
            expect(
                screen.getByRole('option', { name: minDate.getFullYear().toString() })
            ).toBeInTheDocument();
            expect(
                screen.getByRole('option', { name: maxDate.getFullYear().toString() })
            ).toBeInTheDocument();
        });

        it('should set year boundaries on header to default', async () => {
            const user = userEvent.setup();
            const currentYear = new Date().getFullYear();

            await openCalendar(user);

            expect(screen.getByText(currentYear - 100)).toBeInTheDocument();
            expect(screen.getByText(currentYear + 100)).toBeInTheDocument();
        });
    });
});
