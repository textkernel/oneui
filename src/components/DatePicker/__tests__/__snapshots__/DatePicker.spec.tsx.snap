// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker should render correctly 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class=""
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`DatePicker when no min/max date props passed should set min/max date to default 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class=""
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
